File: sdk/python/examples/chat_completion.py
#!/usr/bin/env python3
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------

"""Example: Chat completion using Foundry Local Python SDK.

Demonstrates basic chat completion with the Foundry Local runtime,
including model discovery, loading, and inference.
"""

from foundry_local_sdk import Configuration, FoundryLocalManager

def main():
    # 1. Initialize the SDK
    config = Configuration(app_name="ChatCompletionExample")
    print("Initializing Foundry Local Manager")
    FoundryLocalManager.initialize(config)
    manager = FoundryLocalManager.instance

    # 2. Print available models in the catalog and cache
    models = manager.catalog.list_models()
    print("Available models in catalog:")
    for m in models:
        print(f"  - {m.alias} ({m.id})")

    cached_models = manager.catalog.get_cached_models()
    print("\nCached models:")
    for m in cached_models:
        print(f"  - {m.alias} ({m.id})")

    CACHED_MODEL_ALIAS = "qwen2.5-0.5b"

    # 3. Find a model from the cache (+ download if not cached)
    model = manager.catalog.get_model(CACHED_MODEL_ALIAS)
    if model is None:
        print(f"Model '{CACHED_MODEL_ALIAS}' not found in catalog.")
        print("Available models:")
        for m in manager.catalog.list_models():
            print(f"  - {m.alias} ({m.id})")
        return

    if not model.is_cached:
        print(f"Downloading {model.alias}...")
        model.download(progress_callback=lambda pct: print(f"  {pct:.1f}%", end="\r"))
        print()

    # 4. Load the model
    print(f"Loading {model.alias}...", end="")
    model.load()
    print("loaded!")

    try:
        # 5. Create a chat client and send a message
        client = model.get_chat_client()

        print("\n--- Non-streaming ---")
        response = client.complete_chat(
            messages=[{"role": "user", "content": "What is the capital of France? Reply briefly."}]
        )
        print(f"Response: {response.choices[0].message.content}")

        # 6. Streaming example
        print("\n--- Streaming ---")

        def on_chunk(chunk):
            if chunk.choices and chunk.choices[0].delta and chunk.choices[0].delta.content:
                print(chunk.choices[0].delta.content, end="", flush=True)

        client.complete_streaming_chat(
            [{"role": "user", "content": "Tell me a short joke."}],
            on_chunk,
        )
        print()  # newline after streaming

    except Exception as e:
        print(f"Error during inference: {e}")

    finally:
        # 7. Cleanup
        model.unload()
        print("\nModel unloaded.")


if __name__ == "__main__":
    main()

File: sdk/js/examples/chat-completion.ts
// -------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
// -------------------------------------------------------------------------

import { FoundryLocalManager } from '../src/index.js';

async function main() {
    try {
        // Initialize the Foundry Local SDK
        console.log('Initializing Foundry Local SDK...');
        
        // NOTE: You must update libraryPath to point to your built DLL if not in standard location
        const manager = FoundryLocalManager.create({
            appName: 'FoundryLocalExample',
            serviceEndpoint: 'http://localhost:5000',
            logLevel: 'info'
        });
        console.log('✓ SDK initialized successfully');

        const availableEps = manager.discoverEps();
        console.log(`\nAvailable execution providers: ${availableEps.map((ep) => ep.name).join(', ')}`);

        console.log('\nDownloading and registering execution providers...');
        const downloadResult = manager.downloadAndRegisterEps();
        if (downloadResult.success) {
            console.log('✓ All execution providers registered successfully');
        } else {
            console.log(`⚠️ Some execution providers failed to download and/or register: ${downloadResult.failedEps.join(', ')}`);
        }

        // Explore available models
        console.log('\nFetching available models...');
        const catalog = manager.catalog;
        const models = await catalog.getModels();
        
        console.log(`Found ${models.length} models:`);
        for (const model of models) {
            const variants = model.variants.map((v: any) => v.id).join(', ');
            console.log(`  - ${model.alias} (variants: ${variants})`);
        }

        // Explore cached models
        console.log('\nFetching cached models...');
        const cachedModels = await catalog.getCachedModels();
        console.log(`Found ${cachedModels.length} cached models:`);
        for (const cachedModel of cachedModels) {
            console.log(`  - ${cachedModel.alias}`);
        }

        const modelAlias = 'qwen2.5-0.5b';
        
        // Load the model first
        console.log(`\nLoading model ${modelAlias}...`);
        const modelToLoad = await catalog.getModel(modelAlias);
        if (!modelToLoad) {
            throw new Error(`Model ${modelAlias} not found`);
        }

        await modelToLoad.load();
        console.log('✓ Model loaded');

        // Create chat client
        console.log('\nCreating chat client...');
        const chatClient = modelToLoad.createChatClient();
        
        // Configure chat settings
        chatClient.settings.temperature = 0.7;
        chatClient.settings.maxTokens = 800;
        
        console.log('✓ Chat client created');

        // Example chat completion
        console.log('\nTesting chat completion...');
        const completion = await chatClient.completeChat([
            { role: 'user', content: 'What is the capital of France?' }
        ]);

        console.log('\nChat completion result:');
        console.log(completion.choices[0]?.message?.content);

        // Example streaming completion
        console.log('\nTesting streaming completion...');
        for await (const chunk of chatClient.completeStreamingChat(
            [{ role: 'user', content: 'Write a short poem about programming.' }]
        )) {
            const content = chunk.choices?.[0]?.message?.content;
            if (content) {
                process.stdout.write(content);
            }
        }
        console.log('\n');

        // Model management example
        const model = await catalog.getModel(modelAlias);
        if (model) {
            console.log('\nModel management example:');
            // Already loaded above, but let's check status
            console.log(`Checking model: ${model.id}`);
            console.log(`✓ Model loaded: ${await model.isLoaded()}`);
            
            // Unload when done
            console.log('Unloading model...');
            await model.unload();
            console.log(`✓ Model loaded: ${await model.isLoaded()}`);
        }

        console.log('\n✓ Example completed successfully');

    } catch (error) {
        console.log('Error running example:', error);
        // Print stack trace if available
        if (error instanceof Error && error.stack) {
            console.log(error.stack);
        }
        process.exit(1);
    }
}

// Run the example
main().catch(console.error);

export { main };